ui/ui.js

```diff
@@ -33,163 +33,225 @@ var AGENT_TYPES = {
-    var worker = simlib;
-    if (worker2) {
-        worker = simlib2;
-        worker2 = false;
-    } else {
-        worker2 = true;
-    }
-    worker.postMessage({
-        msg: "statweight", id: id, payload: {
-            iters: iters,
-            dur: dur,
-            numClTargets: numClTargets,
-            gearlist: gearlist,
-            opts: opts,
-            stat: statToMod,
-            modVal: modAmount
-        }
-    });
-}
-
-function packOptions(opt, onComplete) {
-    var id = makeid();
-    simrequests[id] = onComplete
-    simlib.postMessage({
-        msg: "packopt", id: id, payload: {
-            opt: opt,
-        }
-    });
+// This must be kept in sync with the const values in stats.go
+const STAT_IDX = {
+	INT:        0,
+	STAM:       1,
+	SPELL_CRIT: 2,
+	SPELL_HIT:  3,
+	SPELL_DMG:  4,
+	HASTE:      5,
+	MP5:        6,
+	MANA:       7,
+	SPELL_PEN:  8,
+	SPIRIT:     9
+};
+const STATS_LEN = 10;
+
+class SimWorker {
+	constructor() {
+		this.numTasksRunning = 0;
+		this.taskIdsToPromiseFuncs = {};
+		this.worker = new window.Worker('simworker.js');
+
+		let resolveReady = null;
+		this.onReady = new Promise((_resolve, _reject) => {
+			resolveReady = _resolve;
+		});
+
+		this.worker.onmessage = event => {
+			if (event.data.msg == "ready") {
+					this.worker.postMessage({ msg: "setID", payload: "1" });
+					resolveReady();
+			} else if (event.data.msg == "idconfirm") {
+				// Do nothing
+			} else {
+					const id = event.data.id;
+					if (!this.taskIdsToPromiseFuncs[id]) {
+						console.warn("Unrecognized result id: " + id);
+						return;
+					}
+
+					const promiseFuncs = this.taskIdsToPromiseFuncs[id];
+					delete this.taskIdsToPromiseFuncs[id];
+					this.numTasksRunnning--;
+
+					promiseFuncs[0](event.data.payload);
+			}
+		};
+	}
+
+	makeTaskId() {
+			let id = '';
+			const characters = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
+			for (let i = 0; i < 16; i++) {
+					id += characters.charAt(Math.floor(Math.random() * characters.length));
+			}
+			return id;
+	}
+
+	async doApiCall(request) {
+		await this.onReady;
+
+		const taskPromise = new Promise((resolve, reject) => {
+			const id = this.makeTaskId();
+			this.taskIdsToPromiseFuncs[id] = [resolve, reject];
+			this.numTasksRunning++;
+			
+			this.worker.postMessage({
+				msg: "apiCall",
+				id: id,
+				payload: request
+			});
+		});
+		return await taskPromise;
+	}
 }
 
-function makeid() {
-    var result = '';
-    var characters = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
-    var charactersLength = characters.length;
-    for (var i = 0; i < 16; i++) {
-        result += characters.charAt(Math.floor(Math.random() * charactersLength));
-    }
-    return result;
+class WorkerPool {
+	constructor(numWorkers) {
+		this.workers = [];
+		for (let i = 0; i < numWorkers; i++) {
+			this.workers.push(new SimWorker());
+		}
+	}
+
+	getLeastBusyWorker() {
```

Comment:
awesome, this is much better than my hacked version of this function :)

Reply:
Yep, the whole interface with the sim is waaaay nicer now. In fact, we should never need to touch simworker.js OR main_wasm.go ever again.